sinks/grpsink/grpc_stream.go

```diff
@@ -0,0 +1,123 @@
+	"github.com/stripe/veneur/protocol"
+	"github.com/stripe/veneur/sinks"
+	"github.com/stripe/veneur/ssf"
+	"github.com/stripe/veneur/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+)
+
+// GRPCStreamingSpanSink is a sink that streams spans to a configurable target
+// service over gRPC. The sink is only tied to the grpc_sink.proto definition of
+// a SpanSink service, and thus is generic with respect to the specific server
+// it is connecting to.
+type GRPCStreamingSpanSink struct {
+	name        string
+	grpcConn    *grpc.ClientConn
+	sinkClient  SpanSinkClient
+	stream      SpanSink_SendSpansClient
+	stats       *statsd.Client
+	commonTags  map[string]string
+	traceClient *trace.Client
+	log         *logrus.Logger
+}
+
+var _ sinks.SpanSink = &GRPCStreamingSpanSink{}
+
+// NewGRPCStreamingSpanSink creates a sinks.SpanSink that can write to any
+// compliant gRPC server.
+//
+// The target parameter should be of the "host:port"; the name parameter is
+// prepended with "grpc-", and is used when reporting logs in order to permit
+// differentiation between various services.
+//
+// Any grpc.CallOpts that are provided will be used while first establishing the
+// connection to the target server (in grpc.DialContext()).
+func NewGRPCStreamingSpanSink(ctx context.Context, target, name string, commonTags map[string]string, log *logrus.Logger, opts ...grpc.DialOption) (*GRPCStreamingSpanSink, error) {
+	name = "grpc-" + name
+	conn, err := grpc.DialContext(ctx, target, opts...)
+	if err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			"name":   name,
+			"target": target,
+		}).Error("Error establishing connection to gRPC server")
+		return nil, err
+	}
+
+	return &GRPCStreamingSpanSink{
+		grpcConn:   conn,
+		name:       name,
+		sinkClient: NewSpanSinkClient(conn),
+		commonTags: commonTags,
+		log:        log,
+	}, nil
+}
+
+func (gs *GRPCStreamingSpanSink) Start(cl *trace.Client) error {
+	// TODO this API is experimental - perhaps we shouldn't use it yet
+	if !gs.grpcConn.WaitForStateChange(context.TODO(), connectivity.Ready) {
+		err := errors.New("gRPC connection failed to become ready")
+		gs.log.WithError(err).WithFields(logrus.Fields{
+			"name":          gs.name,
+			"current-state": gs.grpcConn.GetState().String(),
+		}).Error("Error establishing connection to gRPC server")
+		return err
+	}
+
+	// TODO How do we allow grpc.CallOption to be injected in here?
+	// TODO Setting this up here (probably) means one stream per lifecycle of this object. Is that the right design?
+	stream, err := gs.sinkClient.SendSpans(context.TODO())
+	if err != nil {
+		gs.log.WithFields(logrus.Fields{
+			"name":          gs.name,
+			logrus.ErrorKey: err,
+		}).Error("Failed to set up a gRPC stream to sink target")
+		return err
+	}
+
+	gs.stream, gs.traceClient = stream, cl
+	return nil
+}
+
+// Name returns this sink's name.
+func (gs *GRPCStreamingSpanSink) Name() string {
+	return gs.name
```

Comment:
Hm, we use the `Name` a ton in metric tags, which means that every grpc sink gets its own timeseries; I guess this is one of those blessing/curse things, but we should definitely think through if we want to time things by grpc sink implementation, or by sink target (currently, we measure only by sink implementation)

Reply:
My thought was that it seems unlikely that the grpc sink gets used directly - instead, i can imagine e.g. a very lightweight falconer sink that just wraps this one. That way we'd still have stability on this property, even though it isn't hardwired into the struct.

Alternatively, we could enforce that things work that way by dropping the `Name()` method entirely, ,causing `GRPCStreamingSpanSink` to not fulfill the interface anymore, and _requiring_ some other struct (e.g. falconer) to embed it in order to have it actually be useful. That smacks of inheritance, but not so much that it wouldn't be idiomatic.

Reply:
So, alternative: what if the expectation is that a value in a configuration file is what sets the actual name, here?